File: go/internal/synthesis/diagram_builder.go
// Package synthesis — DiagramBuilder generates Mermaid diagrams from relationships.
package synthesis

import (
	"fmt"
	"regexp"
	"sort"
	"strings"

	"github.com/unrealandychan/rekipedia/internal/models"
)

var reNodeSafe = regexp.MustCompile(`[^A-Za-z0-9_]`)

// DiagramBuilder generates Mermaid diagrams from relationship data.
type DiagramBuilder struct{}

// NewDiagramBuilder returns a new DiagramBuilder.
func NewDiagramBuilder() *DiagramBuilder { return &DiagramBuilder{} }

// Build generates diagrams from relationships and entry points.
// Returns map[name][2]string where [0]=type, [1]=mermaid content.
func (d *DiagramBuilder) Build(relationships []models.Relationship, entryPoints []string) map[string][2]string {
	out := make(map[string][2]string)
	if len(relationships) == 0 {
		return out
	}

	out["module-graph"] = [2]string{"flowchart", buildModuleGraph(relationships, entryPoints)}

	if ch := buildClassHierarchy(relationships); ch != "" {
		out["class-hierarchy"] = [2]string{"classDiagram", ch}
	}

	return out
}

// ── module graph ──────────────────────────────────────────────────────────────

func buildModuleGraph(rels []models.Relationship, entryPoints []string) string {
	entrySet := make(map[string]bool)
	for _, ep := range entryPoints {
		entrySet[ep] = true
	}

	type edge struct{ from, to, kind string }
	var importEdges, callEdges, inheritEdges []edge
	seenNodes := make(map[string]bool)

	for _, r := range rels {
		from := r.From
		to := r.To
		if from == "" || to == "" {
			continue
		}
		switch r.Kind {
		case models.RelImport:
			// Only include if the target looks like an internal module
			if strings.Contains(to, "/") || strings.Contains(to, ".") || isInternalLookingModule(to) {
				importEdges = append(importEdges, edge{from, to, "import"})
				seenNodes[from] = true
				seenNodes[to] = true
			}
		case models.RelCall:
			callEdges = append(callEdges, edge{from, to, "call"})
			seenNodes[from] = true
			seenNodes[to] = true
		case models.RelInherits:
			inheritEdges = append(inheritEdges, edge{from, to, "inherits"})
			seenNodes[from] = true
			seenNodes[to] = true
		}
	}

	if len(seenNodes) == 0 {
		return "flowchart LR\n  A[No internal relationships detected]"
	}

	var lines []string
	lines = append(lines, "flowchart LR")

	// Node definitions with display labels
	defined := make(map[string]bool)
	sortedNodes := sortedKeys(seenNodes)
	for _, name := range sortedNodes {
		nid := nodeID(name)
		if defined[nid] {
			continue
		}
		defined[nid] = true
		label := nodeLabel(name)
		if entrySet[name] {
			lines = append(lines, fmt.Sprintf("  %s([%q])", nid, label))
		} else {
			lines = append(lines, fmt.Sprintf("  %s[%q]", nid, label))
		}
	}

	// Limit edges per type to keep diagram readable (max 30 each)
	const maxEdges = 30

	for i, e := range importEdges {
		if i >= maxEdges {
			break
		}
		lines = append(lines, fmt.Sprintf("  %s -->|import| %s", nodeID(e.from), nodeID(e.to)))
	}
	for i, e := range callEdges {
		if i >= maxEdges {
			break
		}
		lines = append(lines, fmt.Sprintf("  %s -.->|call| %s", nodeID(e.from), nodeID(e.to)))
	}
	for _, e := range inheritEdges {
		lines = append(lines, fmt.Sprintf("  %s -->|extends| %s", nodeID(e.from), nodeID(e.to)))
	}

	return strings.Join(lines, "\n")
}

// ── class hierarchy ───────────────────────────────────────────────────────────

func buildClassHierarchy(rels []models.Relationship) string {
	var inherits []models.Relationship
	for _, r := range rels {
		if r.Kind == models.RelInherits {
			inherits = append(inherits, r)
		}
	}
	if len(inherits) == 0 {
		return ""
	}

	var lines []string
	lines = append(lines, "classDiagram")
	seen := make(map[string]bool)
	for _, r := range inherits {
		key := r.From + "->" + r.To
		if seen[key] {
			continue
		}
		seen[key] = true
		lines = append(lines, fmt.Sprintf("  %s <|-- %s", safeClassName(r.To), safeClassName(r.From)))
	}

	if len(lines) == 1 {
		return "" // only header, no actual relationships
	}
	return strings.Join(lines, "\n")
}

// ── helpers ───────────────────────────────────────────────────────────────────

func nodeID(name string) string {
	return reNodeSafe.ReplaceAllString(name, "_")
}

func nodeLabel(name string) string {
	parts := strings.Split(name, "/")
	last := parts[len(parts)-1]
	dotParts := strings.Split(last, ".")
	return dotParts[0]
}

func safeClassName(name string) string {
	s := reNodeSafe.ReplaceAllString(name, "_")
	if s == "" {
		return "Unknown"
	}
	return s
}

// isInternalLookingModule heuristically detects internal imports.
// Keeps anything that isn't a known stdlib / popular third-party package.
func isInternalLookingModule(name string) bool {
	external := map[string]bool{
		// Go stdlib
		"os": true, "sys": true, "re": true, "json": true, "math": true,
		"io": true, "fmt": true, "log": true, "time": true, "net": true,
		"http": true, "strings": true, "strconv": true, "errors": true,
		"context": true, "sync": true, "path": true, "sort": true,
		"bytes": true, "bufio": true, "crypto": true, "reflect": true,
		"runtime": true, "testing": true, "flag": true, "encoding": true,
		// Python stdlib
		"typing": true, "pathlib": true, "collections": true, "itertools": true,
		"functools": true, "dataclasses": true, "enum": true, "abc": true,
		"logging": true, "threading": true, "subprocess": true, "asyncio": true,
		"contextlib": true, "copy": true, "hashlib": true, "base64": true,
		"uuid": true, "random": true, "datetime": true, "struct": true,
		// Popular third-party
		"fastapi": true, "pydantic": true, "sqlalchemy": true, "django": true,
		"flask": true, "requests": true, "httpx": true, "aiohttp": true,
		"numpy": true, "pandas": true, "torch": true, "sklearn": true,
		"openai": true, "boto3": true, "celery": true, "redis": true,
		"pytest": true, "click": true, "typer": true, "rich": true,
		"starlette": true, "uvicorn": true, "gunicorn": true, "alembic": true,
		"litellm": true, "faiss": true,
	}
	return !external[name]
}

func sortedKeys(m map[string]bool) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

File: src/rekipedia/synthesis/diagram_builder.py
"""Build Mermaid diagrams from relationship data."""
from __future__ import annotations

import re


class DiagramBuilder:
    """Generate Mermaid diagrams from relationship rows."""

    def build(self, relationships: list, entry_points: list[str] | None = None) -> dict[str, tuple[str, str]]:
        """Return {name: (diagram_type, mermaid_content)}."""
        results: dict[str, tuple[str, str]] = {}
        if not relationships:
            return results
        rows = [_to_dict(r) for r in relationships]
        results["module-graph"] = ("flowchart", _build_module_graph(rows, entry_points=entry_points))
        results["class-hierarchy"] = ("classDiagram", _build_class_hierarchy(rows))
        return results


# ── diagram builders ─────────────────────────────────────────────────

def _build_module_graph(rows: list[dict], entry_points: list[str] | None = None) -> str:
    """flowchart LR showing module relationships with labelled edges and entry point highlights."""
    entry_set = set(entry_points or [])

    # Collect all edges by type
    import_edges: list[tuple[str, str]] = []
    call_edges: list[tuple[str, str]] = []
    inherit_edges: list[tuple[str, str]] = []

    seen_nodes: set[str] = set()

    for r in rows:
        frm = r.get("from_") or r.get("from") or ""
        to = r.get("to") or ""
        kind = r.get("kind") or ""
        if not frm or not to:
            continue
        # Only keep internal relationships — filter out stdlib/well-known external packages
        _EXTERNAL_PREFIXES = (
            "os", "sys", "re", "io", "json", "time", "math", "typing", "pathlib",
            "collections", "itertools", "functools", "dataclasses", "enum", "abc",
            "logging", "threading", "subprocess", "asyncio", "contextlib", "copy",
            "hashlib", "base64", "uuid", "random", "datetime", "string", "struct",
            # popular third-party
            "fastapi", "pydantic", "sqlalchemy", "django", "flask", "requests",
            "httpx", "aiohttp", "numpy", "pandas", "torch", "sklearn", "openai",
            "boto3", "celery", "redis", "pytest", "click", "typer", "rich",
            "starlette", "uvicorn", "gunicorn", "alembic", "litellm", "faiss",
        )
        if kind == "import":
            # Keep if it looks internal: has path separator, has dot-notation, or
            # doesn't start with a known external package name
            is_external = any(to == p or to.startswith(p + ".") or to.startswith(p + "/")
                              for p in _EXTERNAL_PREFIXES)
            if not is_external and to:
                import_edges.append((frm, to))
                seen_nodes.add(frm)
                seen_nodes.add(to)
        elif kind == "call":
            call_edges.append((frm, to))
            seen_nodes.add(frm)
            seen_nodes.add(to)
        elif kind == "inherits":
            inherit_edges.append((frm, to))
            seen_nodes.add(frm)
            seen_nodes.add(to)

    if not seen_nodes:
        return "flowchart LR\n  A[No internal relationships detected]"

    # Build id -> display label mapping
    def node_id(name: str) -> str:
        return re.sub(r"[^A-Za-z0-9_]", "_", name)

    def node_label(name: str) -> str:
        # Use last path component or class name as display label
        label = name.split("/")[-1].split(".")[0]
        return label or name

    lines = ["flowchart LR"]

    # Define nodes with display labels
    defined: set[str] = set()
    for name in sorted(seen_nodes):
        nid = node_id(name)
        if nid not in defined:
            label = node_label(name)
            lines.append(f'  {nid}["{label}"]')
            defined.add(nid)

    lines.append("")

    # Import edges (limit 25)
    for frm, to in import_edges[:25]:
        lines.append(f"  {node_id(frm)} -->|imports| {node_id(to)}")

    # Call edges (limit 15)
    for frm, to in call_edges[:15]:
        lines.append(f"  {node_id(frm)} -.->|calls| {node_id(to)}")

    # Inheritance edges (limit 10)
    for child, parent in inherit_edges[:10]:
        lines.append(f"  {node_id(parent)} <|-- {node_id(child)} : inherits")

    lines.append("")

    # Style entry points in gold
    for ep in (entry_points or []):
        nid = node_id(ep)
        if nid in defined:
            lines.append(f"  style {nid} fill:#f4a700,stroke:#c47d00,color:#000")

    return "\n".join(lines)


def _build_class_hierarchy(rows: list[dict]) -> str:
    """classDiagram showing inheritance relationships."""
    inheritances: list[tuple[str, str]] = []

    for r in rows:
        if r.get("kind") == "inherits":
            frm = r.get("from_") or r.get("from") or ""
            to = r.get("to") or ""
            if frm and to:
                inheritances.append((_sanitise_class(frm), _sanitise_class(to)))

    if not inheritances:
        return "classDiagram\n  note \"No inheritance relationships detected\""

    lines = ["classDiagram"]
    for child, parent in inheritances[:20]:
        lines.append(f"  {parent} <|-- {child}")
    return "\n".join(lines)


# ── helpers ──────────────────────────────────────────────────────────

def _to_dict(row) -> dict:
    if isinstance(row, dict):
        return row
    try:
        return dict(row)
    except Exception:  # noqa: BLE001
        return {}


def _sanitise(name: str) -> str:
    """Convert a file path to a valid Mermaid node id."""
    # Replace non-alphanumeric characters
    s = re.sub(r"[^A-Za-z0-9_]", "_", name)
    # Ensure doesn't start with digit
    if s and s[0].isdigit():
        s = "_" + s
    return s or "unknown"


def _sanitise_class(name: str) -> str:
    """Strip fully-qualified prefix for class names."""
    return re.sub(r"[^A-Za-z0-9_]", "_", name.split(".")[-1]) or "Unknown"
